wp_derive_request_builder/src/generate.rs

```diff
@@ -64,7 +64,23 @@ fn generate_async_request_executor(config: &Config, parsed_enum: &ParsedEnum) ->
             quote! {
                 pub async #fn_signature -> Result<#output_type, #static_wp_api_error_type> {
                     #request_from_request_builder
-                    self.request_executor.execute(std::sync::Arc::new(request)).await?.parse()
+
+                    let cloned_request = request.clone();
+                    let result = self.request_executor.execute(request.into()).await;
+
+                    if let Ok(response) = &result {
+                        if response.status_code == 401 {
+                            if let crate::WpAuthentication::UserAccount { ref login } = self.request_builder.inner.authentication {
+                                let client = crate::login::WpLoginClient::new(self.request_executor.clone());
+                                let api_base_url = self.request_builder.endpoint.api_base_url.clone();
+                                if let Some(request) = client.insert_rest_nonce(&cloned_request, &api_base_url, login).await {
+                                    return self.request_executor.execute(request.into()).await?.parse();
+                                }
+                            }
+                        }
+                    }
+
+                    result?.parse()
```

Comment:
> We want the generated code logic to remain simple and more importantly it shouldn't contain any "magic".

I tried to do that, by having minimal code in here.  Maybe it's because I'm new to Rust, but writing and debugging code in this proc macro is not a super straightforward experience.

These statements are the minimal code to call the `insert_rest_nonce` function. But we can move them all out into a static function if that's easier to read.

> We probably want a more structured retry and/or re-authenticate strategy that clients can opt into or out of.

Interesting. I don't see this new code as retry or re-authenticate though. It's part of the cookie authentication, where we have to get the `nonce` value from the site, and pass it to REST API calls. That is not something clients can opt-out of, if they want to use an actual account username and password to authenticate REST API calls.

> due to the request_executor being cloned, the original executor's authentication won't be updated

The request_executor is `Arc<dyn RequestExecutor>`. I'm not sure what cloning it would do. Does it creat another copy of the actual `trait RequestExecutor` instance or just create a new pointer to the same `trait RequestExecutor` instance? Can `trait RequestExecutor` even be cloned?

BTW, the `Arc<dyn RequestExecturo>` passed to WpApiClient is cloned multiple times at the moment:

```
    impl WpApiClient {
        pub fn new(
            site_url: Arc<ParsedUrl>,
            authentication: WpAuthentication,
            request_executor: Arc<dyn RequestExecutor>,
        ) -> Self {
            let api_base_url: Arc<ApiBaseUrl> = Arc::new(site_url.inner.clone().into());
            let nonce_cache = NonceCache::new(request_executor.clone());
            Self {
                application_passwords: ApplicationPasswordsRequestExecutor::new(
                        api_base_url.clone(),
                        authentication.clone(),
                        request_executor.clone(),
                        &nonce_cache,
                    )
                    .into(),
                plugins: PluginsRequestExecutor::new(
                        api_base_url.clone(),
                        authentication.clone(),
                        request_executor.clone(),
                        &nonce_cache,
                    )
                    .into(),
                post_types: PostTypesRequestExecutor::new(
                        api_base_url.clone(),
                        authentication.clone(),
                        request_executor.clone(),
                        &nonce_cache,
                    )
                    .into(),
                posts: PostsRequestExecutor::new(
                        api_base_url.clone(),
                        authentication.clone(),
                        request_executor.clone(),
                        &nonce_cache,
                    )
                    .into(),
                site_settings: SiteSettingsRequestExecutor::new(
                        api_base_url.clone(),
                        authentication.clone(),
                        request_executor.clone(),
                        &nonce_cache,
                    )
                    .into(),
                users: UsersRequestExecutor::new(
                        api_base_url.clone(),
                        authentication.clone(),
                        request_executor.clone(),
                        &nonce_cache,
                    )
                    .into(),
                wp_site_health_tests: WpSiteHealthTestsRequestExecutor::new(
                        api_base_url.clone(),
                        authentication.clone(),
                        request_executor.clone(),
                        &nonce_cache,
                    )
                    .into(),
                nonce_cache,
            }
        }
    }
```

Reply:
> The request_executor is Arc<dyn RequestExecutor>. I'm not sure what cloning it would do. Does it creat another copy of the actual trait RequestExecutor instance or just create a new pointer to the same trait RequestExecutor instance? Can trait RequestExecutor even be cloned?

Sorry, this is something I saw at the last moment and it got mixed up in my mind with cloning the client 🤦 It's totally fine to clone the `RequestExecutor`.

Having said that, I think the original point is maybe still accurate? I am doubting myself, because I still don't think I fully understand how this type of authentication is supposed to work. I guess maybe it works for all clients if the request executor does cookie retention?

The original requests never contain the `nonce` header, so wouldn't all of them fail at first?

Reply:
> I am doubting myself, because I still don't think I fully understand how this type of authentication is supposed to work.

Regarding how it works, I included a brief summary in this PR. [The official doc](https://developer.wordpress.org/rest-api/using-the-rest-api/authentication/#cookie-authentication) have more details. But I can re-cap it here in the context of this library.

First of all, users have to actually sign in their site. In this PR, that's implemented using a `POST wp-login.php` request. That's pretty much the same as submitting the login form from a web browser. If the login HTTP request is successful, the response would include necessary authentication related cookies for the site.

In this PR, I didn't handle cookies, because I defer that to `RequestExecutor`. Basically `RequestExecutor` needs to have the capability of storing cookies in HTTP responses and using stored cookies for HTTP requests.

However, that's not enough for REST API. It also needs a "WP REST Nonce" thing. This nonce is generated and stored in server to verify incoming REST requests. We can get this nonce using `/wp-admin/admin-ajax.php?action=rest-nonce`. Once we get the nonce, we can put it in HTTP request headers or URL query.

And those two things are what REST API cookies authentication needs.

> The original requests never contain the nonce header, so wouldn't all of them fail at first?

Yes, it would. As I noted in my PR description, we should store the nonce somewhere and re-use it. But that's not implemented in this PR.